File: app/src/main/java/com/cdb96/ncmconverter4a/tag/ID3TagBuilder.java
package com.cdb96.ncmconverter4a.tag;
import static java.nio.charset.StandardCharsets.UTF_16LE;
import com.cdb96.ncmconverter4a.util.LengthUtils;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class ID3TagBuilder
{
    private final ByteArrayOutputStream ID3Header = new ByteArrayOutputStream();

    public void initDefaultTagHeader() throws IOException {
        byte[] defaultHeader = {0x49,0x44,0x33,0x03,0x00,0x00,0x00,0x00,0x00,0x00} ;
        ID3Header.write(defaultHeader);
    }

    public byte[] outputHeader() {
        // ID3v2.3 使用同步安全整数表示总大小
        byte[] sizeBytes = LengthUtils.toSyncSafeIntegerBytes(ID3Header.size());
        byte[] ID3HeaderBytes = ID3Header.toByteArray();
        System.arraycopy(sizeBytes,0,ID3HeaderBytes,6,4);
        ID3Header.reset();
        return ID3HeaderBytes;
    }

    public void addCover(byte[] coverData) throws IOException {
        byte[] mimeTypeBytes = "image/jpeg".getBytes();
        byte[] descriptionBytes = "".getBytes();
        byte textEncoding = 0x00; // ISO-8859-1
        byte pictureType = 0x03;

        int frameSize = 1 + mimeTypeBytes.length + 1 + descriptionBytes.length + 1 + 1 + coverData.length;
        // ID3v2.3 帧大小使用普通32位整数，不是同步安全整数
        byte[] frameSizeBytes = LengthUtils.toBigEndianBytes(frameSize);

        ByteBuffer frame = ByteBuffer.allocate(10 + frameSize);
        //header
        frame.order(ByteOrder.BIG_ENDIAN);
        frame.put("APIC".getBytes()); // 帧ID
        frame.put(frameSizeBytes); // 帧大小 (普通整数)
        frame.putShort((short) 0);

        //body
        frame.put(textEncoding); // 文本编码
        frame.put(mimeTypeBytes); // MIME类型
        frame.put((byte) 0); // MIME类型结束符
        frame.put(pictureType); // 图片类型
        frame.put(descriptionBytes); // 描述
        frame.put((byte) 0); // 描述结束符
        frame.put(coverData); // 图片数据

        ID3Header.write(frame.array());
    }

    public void addTIT2(String title) throws Exception {
        byte[] titleBytes = title.getBytes(UTF_16LE);
        int frameSize = 1 + 2 + titleBytes.length; // BOM加2

        // ID3v2.3 帧大小使用普通32位整数
        byte[] frameSizeBytes = LengthUtils.toBigEndianBytes(frameSize);

        ByteBuffer frame = ByteBuffer.allocate(10 + frameSize);
        //header
        frame.order(ByteOrder.BIG_ENDIAN);
        frame.put("TIT2".getBytes());
        frame.put(frameSizeBytes);
        frame.putShort((short) 0);

        //body
        frame.put((byte) 0x01); // UTF-16 编码
        frame.put((byte) 0xFF);  // BOM for UTF-16LE
        frame.put((byte) 0xFE);
        frame.put(titleBytes);

        ID3Header.write(frame.array());
    }

    public void addTPE1(String artist) throws Exception {
        byte[] artistBytes = artist.getBytes(UTF_16LE);
        int frameSize = 1 + 2 + artistBytes.length; // BOM加2

        byte[] frameSizeBytes = LengthUtils.toBigEndianBytes(frameSize);

        ByteBuffer frame = ByteBuffer.allocate(10 + frameSize);
        frame.order(ByteOrder.BIG_ENDIAN);
        frame.put("TPE1".getBytes());
        frame.put(frameSizeBytes);
        frame.putShort((short) 0);

        frame.put((byte) 0x01); // UTF-16 编码
        frame.put((byte) 0xFF);  // BOM for UTF-16LE
        frame.put((byte) 0xFE);
        frame.put(artistBytes);

        ID3Header.write(frame.array());
    }

    public void addTALB(String album) throws Exception {
        byte[] albumBytes = album.getBytes(UTF_16LE);
        int frameSize = 1 + 2 + albumBytes.length; // BOM加2

        byte[] frameSizeBytes = LengthUtils.toBigEndianBytes(frameSize);

        ByteBuffer frame = ByteBuffer.allocate(10 + frameSize);
        frame.order(ByteOrder.BIG_ENDIAN);
        frame.put("TALB".getBytes());
        frame.put(frameSizeBytes);
        frame.putShort((short) 0);

        frame.put((byte) 0x01);
        frame.put((byte) 0xFF);
        frame.put((byte) 0xFE);
        frame.put(albumBytes);

        ID3Header.write(frame.array());
    }
}

File: composeApp/src/commonMain/kotlin/com/cdb96/ncmconverter4a/tag/ID3TagBuilder.kt
package com.cdb96.ncmconverter4a.tag

import com.cdb96.ncmconverter4a.io.BinaryOutput
import com.cdb96.ncmconverter4a.io.write
import com.cdb96.ncmconverter4a.util.LengthUtils.toSyncSafeIntegerBytes
import com.cdb96.ncmconverter4a.util.LengthUtils.writeIntBE
import com.cdb96.ncmconverter4a.util.LengthUtils.writeShortBE

class ID3TagBuilder {
    private val chunks = mutableListOf<ByteArray>()

    fun initDefaultTagHeader() {
        chunks.add(byteArrayOf(0x49, 0x44, 0x33, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00))
    }

    fun outputHeader(): ByteArray {
        val totalSize = prepareHeader()
        val headerBytes = ByteArray(totalSize).also { arr ->
            var offset = 0
            for (chunk in chunks) { chunk.copyInto(arr, offset); offset += chunk.size }
        }
        chunks.clear()
        return headerBytes
    }

    /** Writes frame headers and their bodies directly, without copying the cover. */
    fun writeTo(output: BinaryOutput) {
        prepareHeader()
        try {
            chunks.forEach { output.write(it) }
        } finally {
            chunks.clear()
        }
    }

    private fun prepareHeader(): Int {
        val totalSize = chunks.sumOf { it.size.toLong() }
        require(totalSize in 10L..0x0FFFFFFFL + 10) { "invalid ID3 tag size: $totalSize" }
        // The sync-safe tag size excludes the ten-byte ID3 header.
        toSyncSafeIntegerBytes(totalSize.toInt() - 10).copyInto(chunks.first(), 6)
        return totalSize.toInt()
    }

    private fun addFrame(id: String, vararg bodies: ByteArray) {
        val header = ByteArray(10)
        id.encodeToByteArray().copyInto(header)
        writeIntBE(header, 4, bodies.sumOf { it.size })
        writeShortBE(header, 8, 0)
        chunks.add(header)
        chunks.addAll(bodies)
    }

    fun addCover(coverData: ByteArray) {
        val mimeTypeBytes = "image/jpeg".encodeToByteArray()
        val descriptionBytes = "".encodeToByteArray()
        val textEncoding: Byte = 0x00  // ISO-8859-1
        val pictureType: Byte = 0x03

        var pos = 0
        val body = ByteArray(1 + mimeTypeBytes.size + 1 + descriptionBytes.size + 1 + 1)
        body[pos++] = textEncoding
        mimeTypeBytes.copyInto(body, pos); pos += mimeTypeBytes.size
        body[pos++] = 0  // MIME终止符
        body[pos++] = pictureType
        descriptionBytes.copyInto(body, pos); pos += descriptionBytes.size
        body[pos++] = 0  // 描述终止符
        addFrame("APIC", body, coverData)
    }

    // UTF-16LE 手动编码 (commonMain 无 toByteArray(Charset))
    private fun String.encodeUtf16LE(): ByteArray {
        val out = ByteArray(length * 2)
        for (i in indices) {
            val ch = this[i].code
            out[i * 2] = ch.toByte()
            out[i * 2 + 1] = (ch shr 8).toByte()
        }
        return out
    }

    fun addTIT2(title: String) {
        val titleBytes = title.encodeUtf16LE()
        var pos = 0
        val body = ByteArray(1 + 2 + titleBytes.size)
        body[pos++] = 0x01  // UTF-16 编码
        body[pos++] = 0xFF.toByte()  // BOM UTF-16LE
        body[pos++] = 0xFE.toByte()
        titleBytes.copyInto(body, pos)
        addFrame("TIT2", body)
    }

    fun addTPE1(artist: String) {
        val artistBytes = artist.encodeUtf16LE()
        var pos = 0
        val body = ByteArray(1 + 2 + artistBytes.size)
        body[pos++] = 0x01
        body[pos++] = 0xFF.toByte()
        body[pos++] = 0xFE.toByte()
        artistBytes.copyInto(body, pos)
        addFrame("TPE1", body)
    }

    fun addTALB(album: String) {
        val albumBytes = album.encodeUtf16LE()
        var pos = 0
        val body = ByteArray(1 + 2 + albumBytes.size)
        body[pos++] = 0x01
        body[pos++] = 0xFF.toByte()
        body[pos++] = 0xFE.toByte()
        albumBytes.copyInto(body, pos)
        addFrame("TALB", body)
    }
}
